src/enum.lisp

```diff
@@ -152,99 +241,94 @@
-        (bit-floor 1))
-    (dolist (pair values)
-      ;; bit-floor rule: find the greatest single-bit int used so far,
-      ;; and store its left-shift
-      (destructuring-bind (symbol &optional
-                           (value (prog1 bit-floor
-                                    (setf bit-floor (ash bit-floor 1)))
-                                  value-p))
-          (ensure-list pair)
-        (check-type symbol symbol)
-        (when value-p
-          (check-type value integer)
-          (when (and (>= value bit-floor) (single-bit-p value))
-            (setf bit-floor (ash value 1))))
-        (if (gethash symbol (symbol-values type))
-            (error "A foreign bitfield cannot contain duplicate symbols: ~S."
-                   symbol)
-            (setf (gethash symbol (symbol-values type)) value))
-        (push symbol (gethash value (value-symbols type)))))
-    type))
+  (multiple-value-bind
+        (base-type keyword-values value-keywords
+                   field-keywords bit-index->keyword)
+      (parse-foreign-enum-like type-name base-type values t)
+    (make-instance 'foreign-bitfield
+                   :name type-name
+                   :actual-type (parse-type base-type)
+                   :keyword-values keyword-values
+                   :value-keywords value-keywords
+                   :field-keywords field-keywords
+                   :bit-index->keyword bit-index->keyword)))
 
 (defmacro defbitfield (name-and-options &body masks)
   "Define an foreign enumerated type."
-  (discard-docstring masks)
-  (destructuring-bind (name &optional (base-type :int))
-      (ensure-list name-and-options)
-    `(eval-when (:compile-toplevel :load-toplevel :execute)
-       (notice-foreign-type
-        ',name (make-foreign-bitfield ',name ',base-type ',masks)))))
+  (%defcenum-like name-and-options masks 'make-foreign-bitfield))
 
 (defun foreign-bitfield-symbol-list (bitfield-type)
   "Return a list of SYMBOLS defined in BITFIELD-TYPE."
-  (hash-keys-to-list (symbol-values (parse-type bitfield-type))))
+  (field-keywords (ensure-parsed-base-type bitfield-type)))
 
 (defun %foreign-bitfield-value (type symbols)
-  (reduce #'logior symbols
-          :key (lambda (symbol)
-                 (check-type symbol symbol)
-                 (or (gethash symbol (symbol-values type))
-                     (error "~S is not a valid symbol for bitfield type ~S."
-                            symbol type)))))
+  (declare (optimize speed))
+  (labels ((process-one (symbol)
+             (check-type symbol symbol)
+             (or (gethash symbol (keyword-values type))
+                 (error "~S is not a valid symbol for bitfield type ~S."
+                        symbol type))))
+    (declare (dynamic-extent #'process-one))
+    (if (consp symbols)
+        (reduce #'logior symbols :key #'process-one)
+        (process-one symbols))))
+
+(defmacro %foreign-bitfield-value/form (type symbols)
```

Comment:
yeah, that's some unholy mixture of a thinko and some artifacts of the previous code. got rid of it altogether in a rebased version...

thanks!